_posts/2020-08-11-relay-pagingation-containers-demystified.markdown

```diff
@@ -0,0 +1,234 @@
+
+I like to always start by [reading the docs](https://relay.dev/docs/en/pagination-container). The `direction` key
+is the direction that we paginate through, either `"forward"` or `"backward"`. `getConnectionFromProps` is a
+function that returns the GraphQL connection, in case the query has more than one. And `query` is used to fetch any
+specific page of results.
+
+Those all makes sense to me, but then we arrive at the real gotchas: `getFragmentVariables` and `getVariables`. The
+docs are helpful, but only if you understand
+[the internals of how Relay works](https://relay.dev/docs/en/runtime-architecture.html). Relay has a sophisticated
+architecture that delivers some really well-performing code, but its abstractions sometimes
+"[leak](https://en.wikipedia.org/wiki/Leaky_abstraction)" and you have to deal with implementation details that
+Relay doesn't really want you to even be aware of (in this case, the Relay store).
+
+So what are these two functions? Let's return to the docs:
+
+- `getFragmentVariables` is used when re-rendering the component, to retrieve the previously-fetch GraphQL response
+  for a certain set of variables.
+- `getVariables` is used when actually fetching another page, and its return value is given to the `query`.
+
+I think of `getFragmentVariables` as kind of cache key for lookup in Relay's internal store. Our implementation of
+`getFragmentVariables` above doesn't really do anything interesting, but a connection that accepted `sort` or
+`filter` parameters would need to return those to avoid lookup collisions when the user changed sort and filter
+options.
+
+Now for `getVariables`, which are the variables used for the `query` later on. It really ought to be named
+`getQueryVariables`, I think. But I digress.
+
+Every implementation of `getFragmentVariables` I could find at Artsy was identical, which makes sense because _that
+is the default implementation_. We shouldn't be defining this option at all! As far as I can tell, Artsy started
+with a few pagination containers that supplied this parameter unnecessarily and it got copy-and-pasted throughout
+our codebases.
+
+After revisiting the docs, I noticed other optional parameters that don't need to be defined either. Let's rewrite
+the call to `createPaginationContainer` to only supply the parameters that are required:
+
+```js
+export ArtistFragmentContainer = createPaginationContainer(Artist, {
+  artist: graphql`
+    fragment Artist_artist on Artist @argumentDefinitions(
+      count: { type: "Int", defaultValue: 10 }
+      cursor: { type: "String" } {
+      name
+      id
+      artworksConnection (first: $count, after: $cursor) @connection(key: "Artist_artworks") {
+        edges {
+          node {
+            title # Now fetch all the artwork data
+            id
+          }
+        }
+      }
+    }
+  `
+  }, {
+  getVariables(props, { count, cursor }, fragmentVariables) {
```

Comment:
**thought (non-blocking):** Oof, man am I glad I investigated this a bit more 😅. Look at this nice and squeaky clean config 😝 https://github.com/artsy/eigen/blob/master/src/lib/Scenes/ViewingRoom/ViewingRoomsList.tsx#L153-L159